### PR TITLE
Provide instrumentation support for lettuce-4.0-4.5.0 sync/async APIs

### PR DESCRIPTION
**Pull request by richardstartin**

### CODE REVIEW

File: dd-java-agent/instrumentation/lettuce-4.0/lettuce-4.0.gradle [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+testSets {
+  latestDepTest {
+    dirName = 'test'
+  }
+}
+
+dependencies {
+  compileOnly group: 'biz.paluch.redis', name: 'lettuce', version: '4.0.Final'
+  main_java8CompileOnly group: 'biz.paluch.redis', name: 'lettuce', version: '4.5.0.Final'
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-java-agent/instrumentation/lettuce-4.0/lettuce-4.0.gradle]: Addressed.

[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-java-agent/instrumentation/lettuce-4.0/lettuce-4.0.gradle]: Addressed.

File: dd-java-agent/instrumentation/lettuce-4.0/lettuce-4.0.gradle [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ext {
+  minJavaVersionForTests = JavaVersion.VERSION_1_8
+  maxJavaVersionForTests = JavaVersion.VERSION_1_8
+}
+
+muzzle {
+  pass {
+    group = "biz.paluch.redis"
+    module = "lettuce"
+    versions = "[4.0.Final,4.5.0.Final]"
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-java-agent/instrumentation/lettuce-4.0/lettuce-4.0.gradle]: Addressed.

File: dd-java-agent/instrumentation/lettuce-4.0/src/test/groovy/LettuceAsyncClientTest.groovy [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+import java.util.function.BiConsumer
+import java.util.function.BiFunction
+import java.util.function.Consumer
+import java.util.function.Function
+
+import com.lambdaworks.redis.RedisConnectionException
+
+import static datadog.trace.instrumentation.lettuce.InstrumentationPoints.AGENT_CRASHING_COMMAND_PREFIX
+
+class LettuceAsyncClientTest extends AgentTestRunner {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/lettuce-4.0/src/test/groovy/LettuceAsyncClientTest.groovy:30]: Is there any logic that can be shared between the sync/async tests?

File: settings.gradle [hunk 4]
```diff
@@ -107,6 +107,7 @@ include ':dd-java-agent:instrumentation:jms'
 include ':dd-java-agent:instrumentation:jsp-2.3'
 include ':dd-java-agent:instrumentation:kafka-clients-0.11'
 include ':dd-java-agent:instrumentation:kafka-streams-0.11'
+include ':dd-java-agent:instrumentation:lettuce-4.0'
```
Comment (tylerbenson) [inline settings.gradle:110]: Unless it's important to disambiguate, we've generally dropped the trailing `.0` from the version number.

File: dd-java-agent/instrumentation/lettuce-4.0/src/main/java8/datadog/trace/instrumentation/lettuce/InstrumentationPoints.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  private static boolean isNonInstrumentingKeyword(ProtocolKeyword keyword) {
+    return keyword == SEGFAULT;
+  }
+
+  // Workaround to keep trace agent from crashing
+  // Currently the commands in AGENT_CRASHING_COMMANDS_WORDS will crash the trace agent and
+  // traces with these commands as the resource name will not be processed by the trace agent
+  // https://github.com/DataDog/datadog-trace-agent/blob/master/quantizer/redis.go#L18 has
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/lettuce-4.0/src/main/java8/datadog/trace/instrumentation/lettuce/InstrumentationPoints.java:108]: This url is now out of date... maybe find the new one?

File: dd-java-agent/instrumentation/lettuce-4.0/src/main/java8/datadog/trace/instrumentation/lettuce/InstrumentationPoints.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+import com.lambdaworks.redis.protocol.ProtocolKeyword;
+import com.lambdaworks.redis.protocol.RedisCommand;
+import datadog.trace.bootstrap.instrumentation.api.AgentScope;
+import datadog.trace.bootstrap.instrumentation.api.AgentSpan;
+
+import java.util.EnumSet;
+import java.util.Set;
+import java.util.concurrent.CancellationException;
+
+public final class InstrumentationPoints {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/lettuce-4.0/src/main/java8/datadog/trace/instrumentation/lettuce/InstrumentationPoints.java:27]: I think I'd prefer if this logic was in either the advice classes or the decorator, but I'm also fine leaving it.

File: dd-java-agent/instrumentation/lettuce-4.0/src/main/java8/datadog/trace/instrumentation/lettuce/InstrumentationPoints.java (hunk 5 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/lettuce-4.0/src/main/java8/datadog/trace/instrumentation/lettuce/InstrumentationPoints.java:108]: I looked for it, but couldn't find it, there seems to have been major refactoring here.

File: settings.gradle (hunk 4 above)
Comment (richardstartin) [reply settings.gradle]: Done.

File: dd-java-agent/instrumentation/lettuce-4.0/src/main/java8/datadog/trace/instrumentation/lettuce/InstrumentationPoints.java (hunk 5 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/lettuce-4.0/src/main/java8/datadog/trace/instrumentation/lettuce/InstrumentationPoints.java]: I have removed the link but will follow up to see if this logic can actually just be removed in another PR.

Comment (tylerbenson) [reply dd-java-agent/instrumentation/lettuce-4.0/src/main/java8/datadog/trace/instrumentation/lettuce/InstrumentationPoints.java]: probably this: https://github.com/DataDog/datadog-trace-agent/blob/c56f4c62b8656b3e8fd2c80058fffd0ec6052374/quantizer/redis.go#L18 but was moved to here: https://github.com/DataDog/datadog-agent/blob/1c76b8381a195a0b0f629011a6225e936fe1d37a/pkg/trace/obfuscate/redis.go

**Pull request merged**